vernac/record.ml

```diff
@@ -83,15 +83,23 @@ let interp_fields_evars env sigma ~ninds ~nparams impls_env nots l =
   in
   sigma, (impls, newfs)
 
+(** FIXME: This is a horrible hack, use a saner heuristic *)
+let max_sort s1 s2 = match s1, s2 with
+| (SProp, SProp) | (Prop, Prop) | (Set, Set) -> s1
+| (SProp, (Prop | Set | Type _ as s)) | ((Prop | Set | Type _) as s, SProp) -> s
+| (Prop, (Set | Type _ as s)) | ((Set | Type _) as s, Prop) -> s
+| (Set, Type u) | (Type u, Set) -> Sorts.sort_of_univ (Univ.Universe.sup Univ.Universe.type0 u)
+| (Type u, Type v) -> Sorts.sort_of_univ (Univ.Universe.sup u v)
+
 let compute_constructor_level evars env l =
```

Comment:
This is basically cominductive.sign_level (this one returns the env but the only user ignores the env)

Reply:
The comment of the commit makes it clear that I *do not want* to expose this to the user. I prefer duplicating it here rather than inducing the false belief that the sup of two sorts exists in general.

Reply:
but compute_constructor_level exists in general

Reply:
WDYM? This does not make sense in general, this is just a heuristic to try to read the mind of the user but there is no reason for this code to make sense in a world where we have various sort hierarchies. Its behaviour with SProp is already dubious, to say the least.

Reply:
It does make sense, it's the sup for the order defined by `Sorts.is_sprop u || UGraph.check_leq_sort (universes env) u ind_univ` (as used in indtyping).
Various sort hierarchies are BS.

Reply:
> Various sort hierarchies are BS.

That's the whole point of the sort-poly feature, FYI.

> it's the sup for the order

This is not an order.

Reply:
How is it not an order?

Reply:
Ah, indeed, I misread your code, it is an order. Still, it doesn't correspond to anything in the semantics of the universe relations. Furthermore, this hack is essentially used as a heuristic for the inference of template poly sorts, and you can't tell me that the current behaviour is sane. Like you write `Foo : Type :=` and now it's in Prop. This will have to go away.

Reply:
It has nothing to do with template, it is not a hack, it has the semantics of Box.

Reply:
This is abusing the fact that the boxing chain SProp < Prop < Type is linear. This won't be the case when we have other sorts, and furthermore even without actual new sorts this is going to be a mess with quality variables because now it will be clearly partial and you'll have to make arbitrary choices. Please don't export more of this crap, this is part of the inference algorithm for inductive types, not anything vaguely related to the typing rules of CIC.

Reply:
Inference as in "I read the mind of the user and make non-canonical choices because I am clever" Ltac-tier crap.